File: Data Structures & Algorithms/evaluate-reverse-polish-notation/submission-2.py
class Solution:

    def solve(self, l,r,o):
        match o:
            case "+":
                return l + r
            case "-":
                return l - r
            case "*":
                return l * r
            case "/":
                return int(l / r)


    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        
        t = {'+','-','*','/'}
        for c in tokens:
            if c in t:
                r = stack.pop()
                l = stack.pop()
                stack.append(self.solve(l,r,c))
            else:
                stack.append(int(c))
        return stack.pop()

File: Data Structures & Algorithms/asteroid-collision/submission-2.rs
impl Solution {
    pub fn asteroid_collision(asteroids: Vec<i32>) -> Vec<i32> {
        let mut stack = vec![];

        for asteroid in asteroids {
            let mut destroyed = false;

            while let Some(&top) = stack.last(){
                // already passed the collision point or both going same direction
                if asteroid > 0 || top < 0 { 
                    break;
                } else if top == asteroid.abs() {
                    stack.pop();
                    destroyed = true;
                    break;
                } else if top < asteroid.abs() {
                    stack.pop();
                } else {
                    destroyed = true;
                    break;
                }
            }
            if !destroyed {
                stack.push(asteroid);
            }
        }
        stack
    }
}
